### PR TITLE
[mini] parse fully qualified method name using mono_method_desc_new

### PR DESCRIPTION
**Pull request by rgarc**
parsing fully qualified method name using a`MonoMethodDesc` struct

### CODE REVIEW

File: mono/mini/mini.h [hunk 1]
```diff
@@ -348,6 +348,7 @@ typedef struct MonoBasicBlock MonoBasicBlock;
 typedef struct MonoSpillInfo MonoSpillInfo;
 
 extern MonoCallSpec *mono_jit_trace_calls;
+extern MonoMethodDesc *mono_stats_method_desc;
```
Comment (CoffeeFlux) [inline mono/mini/mini.h]: What's the thought process behind moving the header the method name lives in?

File: mono/mini/mini.c [hunk 2]
```diff
@@ -4125,6 +4126,12 @@ mono_jit_compile_method_inner (MonoMethod *method, MonoDomain *target_domain, in
 
 	mono_domain_lock (target_domain);
 
+	if (mono_stats_method_desc && mono_method_desc_full_match (mono_stats_method_desc, method)) {
+		g_printf ("Printing JIT stats at method: %s\n", mono_method_get_full_name (method));
+		mono_print_jit_stats ();
+		g_printf ("\n");
```
Comment (CoffeeFlux) [inline mono/mini/mini.c]: This should not be necessary. If you need an extra newline, it goes inside the stats printing function.

File: mono/mini/driver.c [hunk 3]
```diff
@@ -1767,7 +1764,7 @@ mono_jit_parse_options (int argc, char * argv[])
 			mono_enable_counters ();
 		} else if (strncmp (argv [i], "--stats=", 8) == 0) {
 			mono_enable_counters ();
-			set_stats_method_name (argv [i] + 8);
+			mono_stats_method_desc = parse_qualified_method_name(argv [i] + 8);
```
Comment (CoffeeFlux) [inline mono/mini/driver.c]: What happens if `mono_stats_method_desc` is already set to a different value?

File: mono/mini/mini.h (hunk 1 above)
Comment (rgarc) [reply mono/mini/mini.h]: I didn't really think about it. I just put it in the same file as `mono_jit_compile_method_inner`, which is where we use it. I can put it back in `mini-runtime.h`

File: mono/mini/driver.c [hunk 4]
```diff
@@ -1689,20 +1689,6 @@ mono_get_version_info (void)
 
 static gboolean enable_debugging;
 
-static void
-set_stats_method_name (char *name)
-{
-	size_t method_name_len = strlen (name);
-	if (method_name_len == 0) {
```
[the remark this answers is not in the document]
Comment (rgarc) [reply mono/mini/driver.c]: I use [`mono_method_desc_new`](https://github.com/mono/mono/blob/0a96741f9374f27757000a3cfcf100d98bcc3758/mono/metadata/debug-helpers.c#L357) for parsing the method name, and it returns `NULL` if you pass in an empty string. So if you do something like `$ mono --stats= main.exe`, [this branch](https://github.com/rgarc/mono/blob/f14d7e99bb0060cc9f4205f3739d8969f46f2836/mono/mini/driver.c#L1704) will trigger, print a parse error, and exit as it should.

File: mono/mini/mini.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (rgarc) [reply mono/mini/mini.h]: I started with MonoCallSpec, but after playing with `mono_callspec_parse`, I realized it's more general purpose. For starters, if we used callspec, the user would have to invoke the stats command with `M:` like so 
```sh
$ mono --stats='M:Program.Test(int)' main.exe
```
With callspec, no matter what input I gave it, I couldn't get it to properly parse the method name. Furthermore, `mono_callspec_parse` uses `MonoMethodDesc` under the hood when parsing method names ([see here](https://github.com/mono/mono/blob/c4aa3277375f62eeba7d13d862f2c563cb10a285/mono/metadata/callspec.c#L259))

File: mono/mini/driver.c (hunk 3 above)
Comment (rgarc) [reply mono/mini/driver.c]: If `mono_stats_method_desc` is already set to a value that points to valid memory on the heap, then there will be a leak. However, I don't want to arbitrarily free memory, so would the idiomatic solution here be to make `mono_stats_method_desc` static and only accessible through getters and setters?

**Comment by CoffeeFlux** [comment]
Discussed feedback in person.

File: mono/mini/driver.c [hunk 5]
```diff
@@ -1711,6 +1697,23 @@ mono_enable_counters (void)
...
 
+static MonoMethodDesc *
+parse_qualified_method_name (char *method_name)
+{
+	if (strlen (method_name) == 0) {
+		g_printf ("Couldn't parse empty method name.");
+		exit (1);
+	}
+	if (mono_stats_method_desc)
+		g_free (mono_stats_method_desc);
```
Comment (CoffeeFlux) [inline mono/mini/driver.c]: I think this is the wrong place to do the free. This is just the static function you've written to parse in the name - the free needs to be at the same place where you set the value to the newly parsed method. This ensures that you only have to take the lock once, for as short a time as possible, and makes obvious to someone else reading the code that you really are freeing the old value.

**Comment by CoffeeFlux** [comment]
@monojenkins build

**Pull request merged**